### ISSUE TITLE
Automatically retry GET/HEAD/etc. on persistent HTTP client connections if it fails before response is received

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5479]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5479|
|Type|enhancement|
|Created|2012-02-11 20:47:18Z|
|Branch|https://github.com/twisted/twisted/tree/agent-retry-5479|

According to http://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-17:

```
6.1.5. Retrying Requests
   Senders can close the transport connection at any time.  Therefore,
   clients, servers, and proxies MUST be able to recover from
   asynchronous close events.  Client software MAY reopen the transport
   connection and retransmit the aborted sequence of requests without
   user interaction so long as the request sequence is idempotent (see
   Section 6.1.2 of [Part2]).  Non-idempotent request methods or
   sequences MUST NOT be automatically retried, although user agents MAY
   offer a human operator the choice of retrying the request(s).
   Confirmation by user-agent software with semantic understanding of
   the application MAY substitute for user confirmation.  The automatic
   retry SHOULD NOT be repeated if the second sequence of requests
   fails.
```

Apparently GET HEAD OPTIONS TRACE PUT DELETE is actually the list of okay methods, but to ease implementation we might want to limit to the subset that has no body (otherwise we need to figure out retryable body producer support).

<details><summary>Searchable metadata</summary>

```
trac-id__5479 5479
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Twisted_12_1 Twisted-12.1
branch__branches_agent_retry_5479 branches/agent-retry-5479
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1328993238000000 1328993238000000
changetime__1332032648000000 1332032648000000
version__None None
owner__itamar itamar
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [33547]) Branching to 'agent-retry-5479'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Almost done, just need to add a flag that disables it, and document the behavior.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

OK, ready for review (hopefully buildbot will be green by the time someone reviews this):
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/agent-retry-5479

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

I wonder why you chose not to implement this as another `Agent`-wrapper.  Apart from configurability, one of the advantages of the composition-based approach was supposed to be implementation simplicity via orthogonality.  Testing the retry logic would be simpler if it were separate from the actual `Agent` code, and there would be minimal (or no) chance of the retry wrapper breaking any existing functionality of `Agent`.  It would also make the code re-usable for any entirely alternate implementation of the underlying `Agent` type.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_6"></a>|@jyknight commented|
|-|-|


```
<itamar> exarkun: oh, hm. I think there's a reason it wasn't a wrapper
* itamar tries to recreate the thought
<itamar> nope, can't recreate it, so I guess it's probably bogus
<foom> internal wrapper, or user-facing api wrapper?
<itamar> foom: I assume you'd suggest it not be user visible?
<foom> well, it's broken to not have it enabled if you use persistent connections. And since a connection pool is part of the core Agent, so should retries be.
<foom> BTW, it looks like you have it turned on unconditionally, rather than just on non-initial requests.
<foom> also, the retry should always be on a brand new connection, not a cached one.
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Reasons not to make it a wrapper:

  1. As James points out, it really should just be the default for persistent connections. This can be solved by documentation, though.
  1. However, using a brand new connection is basically breaking abstraction if you're doing a wrapper. So it basically seems not worth doing separately from Agent.

Another note from JP: retries should be done on TimeoutError.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

As far as "unconditionally, rather than just on initial queries" - if by this you meant initial request by the user, there's a test to make sure we only retry once. If you mean initial request on a given connection, the HTTPbis document doesn't specify that as a restriction. Moreover, existing persistent connections are more likely to fail (due to server timeouts) than new connections. Is there any reason *not* to do retries on requests on existing persistent connections?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst removed owner|
|-|-|

OK, ready for review - I've addressed two of the review comments (TimeoutError, retries using new connections), and explain above why I don't think I need to address the other two.

Build is running at:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/agent-retry-5479

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> As James points out, it really should just be the default for persistent connections. This can be solved by documentation, though.

James said it's "broken" to not automatically retry when persistent connections are enabled.  I expect that it's true most users won't want a connection pool without this retry logic.  What I'm not sure about is whether it necessarily follows that the implementation of the two must be fused together.

Users probably want redirects to be followed, and cookies to be tracked.  Based on what they actually ask for, I expect many want the client to automatically use a proxy, if one is configured in the environment.  However, we've implemented (or will implement) these things as `Agent` wrappers too, not as part of `Agent` itself.

Leaving the _review_ keyword this time.  If you're not convinced then I guess we'll just do it the way you've done in the branch now.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

As I noted, I'd be OK just having a RetryAgent in the persistent connection documentation so people know they want that. However, there's the other issue I noted: retries break the abstraction boundaries, so wrapping is pointless since we'd bypass the underlying Agent anyway... Unless we make the request() method take an new=False kwarg which would percolate all the way down, which is meh. Do you have a different API alternative?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @itamarst|
|-|-|

So, some discussion on IRC let to an alternate API idea.

The idea to keep in mind is that the retry logic is here to prevent the use of the connection pool from introducing new failures.  These failures might arise because a re-used connection is more likely to have problems than a new connection (due to server-side timeouts, primarily, it seems).

This suggests to me that the retry logic belongs in `HTTPConnectionPool`, not in `Agent`.  `HTTPConnectionPool.getConnection` can return a wrapper that applies the desired retry logic to requests itself.  Since it is built by the connection pool, it can very easily get a brand new connection if it wants one (as it seems to, to use for a retry).  It can also very easily tell if it issued a request over a new connection or a re-used connection (so it can avoid doing retries for requests issued over new connections, as it sounds may be desirable - since errors on such requests are likely not caused by the fact that the connection could have become persistent if the request had succeeded).

I think this means the changes to `Agent` are minimal or nil and we don't need a new _new_ flag anywhere.

This does kick out the `TimeoutError` retry feature, which is fine because it's a somewhat different feature anyway, not really related to the use or disuse of persistent connections.  _That_, it seems, can still be implemented as an `Agent` wrapper on a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst removed owner|
|-|-|

I've refactored the code so retries are done by a wrapper class added by the pool. No more TimeoutError retries, and retries are only done for persistent connections, no more new flag. Buildbot is running:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/agent-retry-5479

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez set owner to @itamarst|
|-|-|

This looks like an accidental change:

{{
-class HTTPParserTestsRFCComplaintDelimeter(_HTTPParserTests, TestCase):
+class HTTParserTestsRFCComplaintDelimeter(_HTTPParserTests, TestCase):
}}

Otherwise, +1.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst set status to `closed`|
|-|-|

(In [33907]) Merge agent-retry-5479: Automatric retries for cached persistent HTTP connections.

Author: itamar
Review: exarkun, wsanchez, jknight
Fixes: [#5479](https://github.com/twisted/twisted/issues/5479)

When using cached persistent HTTP client connections, retry requests in certain situations, since using cached connections greatly increases the chance of failure due to server timeouts.
